### PR TITLE
Automatically set first order_by when distinct set

### PR DESCRIPTION
**Pull request by gjaldon**
Addresses this issue: https://github.com/elixir-lang/ecto/issues/495

My initial implementation for this enhancement but will need to add docs and test it some more.

### CODE REVIEW

File: lib/ecto/adapters/postgres/connection.ex [hunk 1]
```diff
@@ -213,12 +213,16 @@ if Code.ensure_loaded?(Postgrex.Connection) do
       end
     end
 
-    defp order_by(order_bys, sources) do
+    defp order_by([], _distincts, _sources), do: nil
```
Comment (josevalim) [inline lib/ecto/adapters/postgres/connection.ex]: If order_by is empty, we should still generate the order by for distincts.

File: lib/ecto/adapters/postgres/connection.ex [hunk 2]
```diff
@@ -71,7 +71,7 @@ if Code.ensure_loaded?(Postgrex.Connection) do
       where    = where(query.wheres, sources)
       group_by = group_by(query.group_bys, sources)
       having   = having(query.havings, sources)
-      order_by = order_by(query.order_bys, sources)
+      order_by = order_by(query.order_bys, query.distincts, sources)
```
Comment (josevalim) [inline lib/ecto/adapters/postgres/connection.ex]: In master/0.9, distinct is singular. I would recommend you to write the patch against master instead. :)

**Comment by ericmj** [comment]
It would be more efficient to build the SQL for distinct once and reuse that in order_by. Would that be possible?

**Comment by josevalim** [comment]
@ericmj yeah.

File: lib/ecto/adapters/postgres/connection.ex (hunk 2 above)
Comment (gjaldon) [inline lib/ecto/adapters/postgres/connection.ex]: Lol didn't realize I was working off of an older version!

**Comment by gjaldon** [comment]
@josevalim @ericmj are the latest changes okay?

File: lib/ecto/adapters/postgres/connection.ex [hunk 3]
```diff
@@ -248,16 +253,21 @@ if Code.ensure_loaded?(Postgrex.Connection) do
...
         Enum.map_join(order_bys, ", ", fn
           %QueryExpr{expr: expr} ->
             Enum.map_join(expr, ", ", &order_by_expr(&1, sources))
         end)
 
-      case exprs do
-        "" -> nil
-        _  -> "ORDER BY " <> exprs
+      case {distinct_exprs, exprs, distinct} do
+        {_, "", _} -> nil
```
Comment (josevalim) [inline lib/ecto/adapters/postgres/connection.ex]: We should add the distinct_exprs to order by even if there is no order_by, no?

Comment (josevalim) [inline lib/ecto/adapters/postgres/connection.ex]: @ericmj said no. :)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by gjaldon** [comment]
Thanks, guys! :+1:

**Pull request merged**